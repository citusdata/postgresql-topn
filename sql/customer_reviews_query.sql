-- Create table to insert summaries.
create table popular_products
(
  review_summary jsonb,
  year double precision,
  month double precision
);


set topn.number_of_counters to 1000;
-- Create different summaries by grouping the reviews according to their year and month.
insert into
  popular_products(review_summary, year, month )
select
  topn_add_agg(product_id),
  extract(year from review_date) as year,
  extract(month from review_date) as month
from
  customer_reviews
group by
  year,
  month;

-- Create another table for the merged results.
create table
  overall_result(merged_summary jsonb);

-- Let's merge the summaries for the overall result and check top-20 items.
insert into
  overall_result(merged_summary)
select
  topn_union_agg(review_summary)
from
  popular_products;

select
  (topn(merged_summary, 20)).*
from
  overall_result
order by 2 DESC, 1;


-- Test window functions
CREATE TABLE daily_populars
(
  date DATE,
  agg_data JSONB
);

INSERT INTO daily_populars
  SELECT
    date_trunc('day', review_date),
    topn_add_agg(product_id)
  FROM 
    customer_reviews
  GROUP BY 
    1;

SELECT 
  date, 
  topn_union_agg(agg_data) OVER seven_days 
FROM 
  daily_populars 
WINDOW 
  seven_days AS (ORDER BY date ASC ROWS 6 PRECEDING)
ORDER BY 
  1
LIMIT 5;
