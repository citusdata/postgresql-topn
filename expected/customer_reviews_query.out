-- Create table to insert summaries.
create table popular_products
(
  review_summary jsonb,
  year double precision,
  month double precision
);
set topn.number_of_counters to 1000;
-- Create different summaries by grouping the reviews according to their year and month.
insert into
  popular_products(review_summary, year, month )
select
  topn_add_agg(product_id),
  extract(year from review_date) as year,
  extract(month from review_date) as month
from
  customer_reviews
group by
  year,
  month;
-- Create another table for the merged results.
create table
  overall_result(merged_summary jsonb);
-- Let's merge the summaries for the overall result and check top-20 items.
insert into
  overall_result(merged_summary)
select
  topn_union_agg(review_summary)
from
  popular_products;
select
  (topn(merged_summary, 20)).*
from
  overall_result
order by 2 DESC, 1;
    item    | frequency 
------------+-----------
 0671003755 |       576
 0613033205 |       575
 0671014730 |       572
 0743527550 |       572
 0375700757 |       555
 0871136791 |       555
 0679460691 |       551
 0375402926 |       550
 0613221583 |       536
 0812550293 |       536
 1575110458 |       536
 1590073983 |       536
 1590073991 |       536
 0375402675 |       524
 0375403477 |       524
 0375703241 |       524
 0399143904 |       524
 0425170349 |       524
 0613222407 |       524
 0765342294 |       510
(20 rows)

